File: android/src/main/kotlin/com/engagepop/flutter/EngagepopPlugin.kt
package com.engagepop.flutter

import android.content.Context
import android.os.Handler
import android.os.Looper
import com.engagepop.EngagePop
import com.engagepop.EngagePopConfig
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/** Flutter bridge over the native EngagePop Android SDK. */
class EngagepopPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {
    private lateinit var context: Context
    private lateinit var channel: MethodChannel
    private lateinit var deepLinkEvents: EventChannel
    private lateinit var inboxEvents: EventChannel
    private val main = Handler(Looper.getMainLooper())

    private var deepLinkSink: EventChannel.EventSink? = null
    private var inboxSink: EventChannel.EventSink? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, "engagepop")
        channel.setMethodCallHandler(this)

        deepLinkEvents = EventChannel(binding.binaryMessenger, "engagepop/deeplinks")
        deepLinkEvents.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) { deepLinkSink = events }
            override fun onCancel(arguments: Any?) { deepLinkSink = null }
        })

        inboxEvents = EventChannel(binding.binaryMessenger, "engagepop/inbox")
        inboxEvents.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) { inboxSink = events }
            override fun onCancel(arguments: Any?) { inboxSink = null }
        })
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        deepLinkEvents.setStreamHandler(null)
        inboxEvents.setStreamHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "configure" -> { configure(call); result.success(null) }
            "requestPushPermission" -> result.success(true) // needs an Activity — see README
            "identify" -> { EngagePop.identify(stringMap(call.argument("attributes"))); result.success(null) }
            "track" -> {
                EngagePop.track(call.argument<String>("event") ?: "", stringMap(call.argument("properties")))
                result.success(null)
            }
            "convert" -> {
                val value = call.argument<Double>("value") ?: 0.0
                val id = call.argument<Int>("campaignId")?.toLong()?.takeIf { it > 0 }
                EngagePop.convert(value, call.argument("order"), id)
                result.success(null)
            }
            "reset" -> { EngagePop.reset(); result.success(null) }
            "refreshInAppMessages" -> { EngagePop.refreshInAppMessages(); result.success(null) }

            // Inbox
            "getInbox" -> {
                val list = EngagePop.inbox?.messages()?.map { m ->
                    mapOf(
                        "id" to m.id, "title" to m.title, "body" to m.body,
                        "url" to m.url, "receivedAt" to m.receivedAt, "read" to m.read,
                    )
                } ?: emptyList<Map<String, Any?>>()
                result.success(list)
            }
            "unreadCount" -> result.success(EngagePop.inbox?.unreadCount() ?: 0)
            "markRead" -> { EngagePop.inbox?.markRead(call.argument<String>("id") ?: ""); result.success(null) }
            "markAllRead" -> { EngagePop.inbox?.markAllRead(); result.success(null) }
            "removeMessage" -> { EngagePop.inbox?.remove(call.argument<String>("id") ?: ""); result.success(null) }
            "clearInbox" -> { EngagePop.inbox?.clear(); result.success(null) }

            else -> result.notImplemented()
        }
    }

    private fun configure(call: MethodCall) {
        val site = call.argument<String>("siteKey") ?: ""
        val app = call.argument<String>("appKey") ?: ""
        val base = call.argument<String>("apiBaseUrl") ?: "https://edge.engagepop.com"
        val debug = call.argument<Boolean>("debugLogging") ?: false
        val autoShow = call.argument<Boolean>("autoShowInAppMessages") ?: true
        EngagePop.configure(context, EngagePopConfig(site, app, base, debug, autoShow))
        // Event sinks must be touched on the main thread.
        EngagePop.deepLinkHandler = { url -> main.post { deepLinkSink?.success(url) } }
        EngagePop.inbox?.onChanged = { main.post { inboxSink?.success(null) } }
    }

    @Suppress("UNCHECKED_CAST")
    private fun stringMap(map: Map<String, Any?>?): Map<String, String> {
        if (map == null) return emptyMap()
        val out = HashMap<String, String>()
        for ((k, v) in map) if (v is String) out[k] = v
        return out
    }
}

File: ios/Classes/EngagepopPlugin.swift
import Flutter
import UIKit
import EngagePop

/// Flutter bridge over the native EngagePop iOS SDK.
public class EngagepopPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
    private var deepLinkSink: FlutterEventSink?
    private let inboxStream = EngagePopStreamHandler()

    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = EngagepopPlugin()
        let channel = FlutterMethodChannel(name: "engagepop", binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(instance, channel: channel)
        let events = FlutterEventChannel(name: "engagepop/deeplinks", binaryMessenger: registrar.messenger())
        events.setStreamHandler(instance)
        let inbox = FlutterEventChannel(name: "engagepop/inbox", binaryMessenger: registrar.messenger())
        inbox.setStreamHandler(instance.inboxStream)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [:]
        switch call.method {
        case "configure":
            configure(args)
            result(nil)
        case "requestPushPermission":
            EngagePop.requestPushAuthorization { granted in result(granted) }
        case "identify":
            EngagePop.identify(args["attributes"] as? [String: String] ?? [:])
            result(nil)
        case "track":
            EngagePop.track(args["event"] as? String ?? "", properties: args["properties"] as? [String: String])
            result(nil)
        case "convert":
            let value = args["value"] as? Double ?? 0
            let id = (args["campaignId"] as? Int).flatMap { $0 > 0 ? Int64($0) : nil }
            EngagePop.convert(value: value, order: args["order"] as? String, campaignID: id)
            result(nil)
        case "reset":
            EngagePop.reset()
            result(nil)
        case "refreshInAppMessages":
            EngagePop.refreshInAppMessages()
            result(nil)
        case "getInbox":
            let messages = EngagePop.shared.inbox?.messages ?? []
            result(messages.map { [
                "id": $0.id, "title": $0.title, "body": $0.body,
                "url": $0.url as Any, "receivedAt": $0.receivedAt.timeIntervalSince1970, "read": $0.read,
            ] })
        case "unreadCount":
            result(EngagePop.shared.inbox?.unreadCount ?? 0)
        case "markRead":
            if let id = args["id"] as? String { EngagePop.shared.inbox?.markRead(id) }
            result(nil)
        case "markAllRead":
            EngagePop.shared.inbox?.markAllRead()
            result(nil)
        case "removeMessage":
            if let id = args["id"] as? String { EngagePop.shared.inbox?.remove(id) }
            result(nil)
        case "clearInbox":
            EngagePop.shared.inbox?.clear()
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func configure(_ args: [String: Any]) {
        let site = args["siteKey"] as? String ?? ""
        let app = args["appKey"] as? String ?? ""
        let base = (args["apiBaseUrl"] as? String).flatMap { URL(string: $0) }
            ?? URL(string: "https://edge.engagepop.com")!
        let debug = args["debugLogging"] as? Bool ?? false
        let autoShow = args["autoShowInAppMessages"] as? Bool ?? true
        EngagePop.configure(EngagePopConfig(
            siteKey: site, appKey: app, apiBaseURL: base, debugLogging: debug, autoShowInAppMessages: autoShow
        ))
        EngagePop.shared.deepLinkHandler = { [weak self] url in
            self?.deepLinkSink?(url.absoluteString)
        }
        NotificationCenter.default.addObserver(
            forName: Inbox.didChangeNotification, object: nil, queue: .main
        ) { [weak self] _ in self?.inboxStream.sink?(nil) }
    }

    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        deepLinkSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        deepLinkSink = nil
        return nil
    }
}

/// A tiny stream handler that just holds a sink — used for the inbox-change
/// event channel (the plugin itself handles the deep-link channel).
final class EngagePopStreamHandler: NSObject, FlutterStreamHandler {
    var sink: FlutterEventSink?
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        sink = events
        return nil
    }
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        sink = nil
        return nil
    }
}
